src/Control/Monad/Catch.hs

```diff
@@ -653,44 +796,61 @@ catches a hs = a `catch` handler
 
 -- | Run an action only if an exception is thrown in the main action. The
 -- exception is not caught, simply rethrown.
+--
+-- /NOTE/ The action is only run if an /exception/ is thrown. If the monad
+-- supports other ways of aborting the computation, the action won't run if
+-- those other kinds of errors are thrown. See 'onError'.
 onException :: MonadCatch m => m a -> m b -> m a
 onException action handler = action `catchAll` \e -> handler >> throwM e
 
+-- | Run an action only if an error is thrown in the main action. Unlike
+-- 'onException', this works with every kind of error, not just exceptions. For
+-- example, if @f@ is an 'ExceptT' computation which aborts with a 'Left', the
+-- computation @onError f g@ will execute @g@, while @onException f g@ will not.
+--
+-- For monads which, unlike 'ExceptT' and 'MaybeT', only have a single exit
```

Comment:
The phrasing of this sentence is a little awkward. I would prefer stating it like:

"This distinction is only meaningful for monads which have multiple exit points, such as 'Except' and 'MaybeT'. For monads that only have a single exit point, there is no difference between 'onException' and 'onError', except that 'onError' has a more constrained type."